core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go

```diff
@@ -33,9 +36,46 @@ type fetchedBlock struct {
 	visited []fetchedLog
 }
 
+func (currentBlock *fetchedBlock) Append(lggr logger.Logger, fl fetchedLog, maxBlockLogs, maxUpkeepLogs int) (fetchedLog, bool) {
+	has, upkeepLogs := currentBlock.has(fl.upkeepID, fl.log)
+	if has {
+		// Skipping known logs
+		return fetchedLog{}, false
+	}
+	// lggr.Debugw("Adding log", "i", i, "blockBlock", currentBlock.blockNumber, "logBlock", log.BlockNumber, "id", id)
+	currentBlock.logs = append(currentBlock.logs, fl)
+
+	// drop logs if we reached limits.
+	if upkeepLogs+1 > maxUpkeepLogs {
+		// in case we have logs overflow for a particular upkeep, we drop a log of that upkeep,
+		// based on shared, random (per block) order of the logs in the block.
+		currentBlock.Sort()
+		var dropped fetchedLog
+		currentLogs := make([]fetchedLog, 0, len(currentBlock.logs)-1)
+		for _, l := range currentBlock.logs {
+			if dropped.upkeepID == nil && l.upkeepID.Cmp(fl.upkeepID) == 0 {
+				dropped = l
+				continue
+			}
+			currentLogs = append(currentLogs, l)
+		}
+		currentBlock.logs = currentLogs
+		return dropped, true
+	} else if len(currentBlock.logs) > maxBlockLogs {
```

Comment:
should this be `len(currentBlock.logs)` + `len(currentBlock.visited)` ? (We account for both in per upkeep limit)